odoo/addons/base/models/ir_qweb.py

```diff
@@ -224,6 +286,55 @@ def _compile_directive_call_assets(self, el, options, indent):
 
     # method called by computing code
 
+    def _start_profiling(self, directive, xpath, options):
+        data = [None, None]
+
+        if self.env.context.get('profile_execution_context_qweb'):
+            # Add some context on thread at current call stack level.
+            # This context stored by collector beside stack and is used by
+            # Speedscope to add a level to the stack with this QWeb directive
+            # information.
+            current_thread = threading.current_thread()
+            previous_context = getattr(current_thread, 'exec_context', ())
+            context = {'directive': directive, 'xpath': xpath}
+            current_thread.exec_context = previous_context + ((tools.profiler.stack_size(), context),)
+            data[0] = previous_context
+
+        if 'log_profiling_hierarchy' in self.env.context:
+            # Add profiling data used by ProfilingQwebView widget and display the
+            # xml with the duration and number of sql of every qweb directives.
+            if not options.get('ref') in self.env.context['log_profiling_arch']:
+                self.env.context['log_profiling_arch'][options.get('ref')] = options.get('document')
+            profile_data = {
+                'view_id': options.get('ref'),
+                'xpath': xpath,
+                'directive': directive,
+                'delay': 0,
+                'query': 0,
+                'now': (time(), self.env.cr.sql_log_count),
+                'children_delay': 0,
+                'children_query': 0,
+            }
+            self.env.context['log_profiling_results'].append(profile_data)
+            self.env.context['log_profiling_hierarchy'].append(profile_data)
+            data[1] = profile_data
+
+        return data
+
+    def _stop_profiling(self, previous_value):
+        previous_context, profile_data = previous_value
+
+        if 'log_profiling_hierarchy' in self.env.context:
+            profile_data['delay'] = time() - profile_data['now'][0] - profile_data['children_delay']
+            profile_data['query'] = self.env.cr.sql_log_count - profile_data['now'][1] - profile_data['children_query']
+            for data in self.env.context['log_profiling_hierarchy']:
+                data['children_delay'] += profile_data['delay']
+                data['children_query'] += profile_data['query']
+
+        if self.env.context.get('profile_execution_context_qweb'):
+            current_thread = threading.current_thread()
+            current_thread.exec_context = previous_context
+
```

Comment:
I wonder if this logic couldn't be hidden in the profiler?

I like the idea that the profiler is almost invisible in the original code, juste call the hooks